### ISSUE TITLE
Clarification on Use Cases/Limitations

### ISSUE THREAD

**Issue by impatient** [arturbosch/detekt#1300]
Trying to find out if this is a valid case for detekt or something we should just change our code to make impossible.

Use case:  When a function has a parameter of type OptionalType, if any child call expression with the same parameter type is not populated warn the user.      

Is this a suitable case for Detekt?  If so, how does one get the function + valid parameter types of a call expression.  I'll gladly submit a PR with documentation clarifications if it makes sense.

Example:  

```kotlin
class Surround(val inner:InnerCall) {

  fun error(x : OptionalType?) {
    // would throw lint error because an argument of type OptionalType is not passed  
    inner.innerCall()
  }

  fun proper(x : OptionalType?) {
    // valid, since value is passed to call
    inner.innerCall(x)
  }

  fun alsoValid() {
    // valid since parent function doesn't have OptionalType argument 
    inner.innerCall()
 }


}

class InnerCall {
  fun innerCall(x : OptionalType? = null) {
   println(x)
  }
}
```

Thanks!

**Comment by schalkms** [comment]
For me this looks like a use case for [custom rules](https://arturbosch.github.io/detekt/extensions.html#custom-rulesets)

**Comment by impatient** [comment]
Good to hear.  That's what I was looking at.  Are there any existing rules that grab type information.  I wasn't seeing any examples there.

Am I going about this the wrong way if I do a Class visitor to identify functions with that signature and stare them in a map, or will I be able dynamically determine called function signatures at the point of the function call?

**Comment by schalkms** [comment]
At the moment there's no type solving in detekt. I hope this will be introduced soon.

**Comment by impatient** [comment]
Updates per discussion in slack:  

Useful context for similar concept of unused parameters.  https://github.com/arturbosch/detekt/blob/master/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt#L229

**Looks like there is no way to preprocess the data before the rules are run as it looks like processors are run out of band.  Which makes storing a lookup table for external function impossible**

https://github.com/arturbosch/detekt/blob/master/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/DetektFacade.kt#L50
 
@schalkms If you don't mind confirming the above that there is no way to force pre-processing the files, I'll resign myself to waiting for the BindingContext.  Thanks again.
